shared/utils/typedefs-list.hpp

```diff
@@ -163,49 +163,376 @@ struct ListWrapper {
         }
     }
 
-    Ptr operator ->() noexcept {
+    Ptr operator->() noexcept {
         return ptr;
     }
 
-    Ptr const operator ->() const noexcept {
+    Ptr const operator->() const noexcept {
         return ptr;
     }
 
     constexpr ListWrapper& operator=(Ptr const& ptr) {
         this->ptr = ptr;
     }
+    constexpr ListWrapper& operator=(Ptr&& ptr) {
+        this->ptr = ptr;
+    }
 
     // method to create a new list easily
-    template<il2cpp_utils::CreationType creationType = il2cpp_utils::CreationType::Temporary>
+    template <il2cpp_utils::CreationType creationType = il2cpp_utils::CreationType::Temporary>
     static ListWrapper<T, Ptr> New() {
         il2cpp_functions::Init();
         auto ls = il2cpp_utils::New<Ptr, creationType>();
         if (!ls) throw ListException(nullptr, "Could not create list!");
 
-        return {*ls};
+        return { *ls };
     }
 
     // method to create a new list easily
-    template<typename U, il2cpp_utils::CreationType creationType = il2cpp_utils::CreationType::Temporary>
+    template <typename U, il2cpp_utils::CreationType creationType = il2cpp_utils::CreationType::Temporary>
     requires(std::is_convertible_v<U, T>)
     static ListWrapper<T, Ptr> New(std::initializer_list<U> values) {
         il2cpp_functions::Init();
         auto ls = il2cpp_utils::New<Ptr, creationType>();
         if (!ls) throw ListException(nullptr, "Could not create list!");
 
-        ls->EnsureCapacity(values.size());
-        for (auto& v : values) ls->Add(v);
+        ListWrapper<T, Ptr> lsWrap = { *ls };
+
+        lsWrap.insert_range(values);
+
+        return lsWrap;
+    }
+
+    /**
+     * @brief System.Collections.Generic::List<T>.IndexOf(T item)
+     *
+     * https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.list-1.indexof?view=net-8.0
+     * @param index
+     * @param item
+     */
+    std::optional<uint_t> index_of(T item) {
+        auto start = this->begin();
+        auto end = this->begin() + this->size();
+        auto it = std::find(start, end, item);
+
+        if (it == end) return std::nullopt;
+
+        return std::distance(start, it);
+    }
+
+    template <typename F>
+    std::optional<T> First(F&& func) {
```

Comment:
the `First` method as C# method name should do a throw on failure to find, and just return T
same goes for Last

if you want something with an optional as return, please name it as a C++ method (lowercase)
I wouldn't mind if the C# method then used that c++ method tho

also for the `OrDefault` method, add a requires clause as so: `...OrDefault() requires(std::is_default_constructible_v<T>) { ... ` 
and maybe construct it using T{} instead of T()

Reply:
Can't you just voluntary throw exceptions with .value()? Defaulting to exceptional behavior doesn't seem like a great idea since it's technically not an error, just no element found.

Reply:
it's more that ArrayW works that way, and FirstOrDefault is supposed to be the non-exception alternative to using straight up First. putting in the layer of nullability with optional is just not how C# works in that regard either so I think we should aim to emulate that as much as we can